worlds/witness/__init__.py

```diff
@@ -150,27 +133,108 @@ def generate_basic(self):
             location_obj = self.multiworld.get_location(event_location, self.player)
             location_obj.place_locked_item(item_obj)
 
-        self.multiworld.itempool += pool
+        # Find out how much empty space there is for junk items. -1 for the "Town Pet the Dog" check
+        itempool_difference = len(self.locat.CHECK_LOCATION_TABLE) - len(self.locat.EVENT_LOCATION_TABLE) - 1
+        itempool_difference -= len(pool)
 
-    def pre_fill(self):
-        # Put good item on first check if there are any of the designated "good items" in the pool
+        # Place two locked items: Good symbol on Tutorial Gate Open, and a Puzzle Skip on "Town Pet the Dog"
         good_items_in_the_game = []
+        plandoed_items = set()
+
+        for v in self.multiworld.plando_items[self.player]:
+            if v.get("from_pool", True):
+                plandoed_items.update({self.items_by_name[i] for i in v.get("items", dict()).keys()
+                                       if i in self.items_by_name})
+                if "item" in v and v["item"] in self.items_by_name:
+                    plandoed_items.add(self.items_by_name[v["item"]])
 
         for symbol in self.items.GOOD_ITEMS:
             item = self.items_by_name[symbol]
-            if item in self.multiworld.itempool: # Only do this if the item is still in item pool (e.g. after plando)
+            if item in pool and item not in plandoed_items:
+                # for now, any item that is mentioned in any plando option, even if it's a list of items, is ineligible.
+                # Hopefully, in the future, plando gets resolved before create_items.
+                # I could also partially resolve lists myself, but this could introduce errors if not done carefully.
```

Comment:
yeah, i think item "fixup" for plando with `from_pool: false` is something that needs a generic solution and not be something that each world should have to work around.

A year ago or so I suggested that we should maybe allow worlds to submit more items than locations, and let core handle/balance it. That way you could just put triforce pieces into the pool instead of replacing specific items. Currently doing so could remove Player2's filler to place Player1's triforce piece, i believe, which is not good. But with from_pool: false this is on the user at least.

I'll try to properly test and merge this tonight.

Reply:
One thing I will say seems silly:

If there is one more item than locations, then one item goes unplaced. In Witness' case, this *always seems to be a Puzzle Skip*, which is a useful / never-exclude / dark blue

It would make more sense to me if junk and traps got dropped before usefuls to make room.

Reply:
My ideal solution would be a flag in the Item class that's called "removable" or something. If it is set to true, that signals to the generator that this item can be safely removed to make room for other items if the amount of items in the MW exceeds the amount of locations. But obviously this is one more thing for world devs to know and worry about.

Reply:
i think it should place non-junk first, so it should always drop junk.

Reply:
oh, nevermind. that is only true for advancement, not for useful